docs/en/platform-engineers/advanced-cue.md

```diff
@@ -0,0 +1,697 @@
+regard the element type of the array list will always be object, and compare the object field with the specified key name.
+If the value of the key name equal, then the patch data will merge into that, if no equal found, the patch will append into the array list.
+
+The example of strategy patch trait will like below:
+ 
+```yaml
+apiVersion: core.oam.dev/v1alpha2
+kind: TraitDefinition
+metadata:
+  annotations:
+    definition.oam.dev/description: "add sidecar to the app"
+  name: sidecar
+spec:
+  appliesToWorkloads:
+    - webservice
+    - worker
+  extension:
+    template: |-
+      patch: {
+         // +patchKey=name
+         spec: template: spec: containers: [parameter]
+      }
+      parameter: {
+         name: string
+         image: string
+         command?: [...string]
+      }
+```
+
+The patchKey is `name` which represents the container name in this example. In this case, if the workload already has
+a container with the same name of this `sidecar` trait, it will be a merge operation. If the workload don't have the container
+with same name, it will be a sidecar container append into the `spec.template.spec.containers` array list.
+
+### Patch works with output
+
+Patch can also work with output, if patch and output both exist in one trait, the patch part will execute first and then 
+the output object will be rendered out. 
+
+```yaml
+apiVersion: core.oam.dev/v1alpha2
+kind: TraitDefinition
+metadata:
+  annotations:
+    definition.oam.dev/description: "service the app"
+  name: kservice
+spec:
+  appliesToWorkloads:
+    - webservice
+    - worker
+  definitionRef:
+    name: services
+  extension:
+    template: |-
+      patch: {spec: template: metadata: labels: app: context.name}
+      output: {
+        apiVersion: "v1"
+        kind: "Service"
+        metadata: name: context.name
+        spec: {
+          selector:  app: context.name
+          ports: [
+            for k, v in parameter.http {
+              port: v
+              targetPort: v
+            }
+          ]
+        }
+      }
+      parameter: {
+        http: [string]: int
+      }
+```
+
+## Processing Trait
+
+A KubeVela trait can also help you to do some processing job. Currently, we have supported http request.
+
+The keyword is `processing`, inside the `processing`, there are two keywords `output` and `http`.
+
+You can define http request `method`, `url`, `body`, `header` and `trailer` in the `http` section.
+KubeVela will send a request using this information, the requested server shall output a **json result**.
+
+The `output` section will used to match with the `json result`, correlate fields by name will be automatically filled into it.
+Then you can use the requested data from `processing.output` into `patch` or `output/outputs`.
+
+Below is an example:
+
+```yaml
+apiVersion: core.oam.dev/v1alpha1
+kind: Trait
+metadata:
+  name: auth-service
+spec:
+  template: |
```

Comment:
this should be a TraitDefinition, right?

Reply:
maybe [this test case](https://github.com/oam-dev/kubevela/blob/9173ff828f289f0f4a4ea5b580d4b84aaf87c818/pkg/controller/core.oam.dev/v1alpha2/application/application_controller_test.go#L830) can help here.

Reply:
yes, that's a mistake, good catch!